Add scripts for creating and destroying assessment environments

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds scripts for creating and destroying assessment environments.

## 💭 Motivation and context ##

This is necessary because the CloudWatch alarm resources associated with the EC2 instances are created using `for_each` expressions that are dynamic-ish, and `terraform` does not handle this very well.

The NOM folks desperately need these scripts in hand to perform their job function.

## 🧪 Testing ##

I was able to use these scripts to successfully create one and destroy several assessment environments.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

jsf9k committed:
Add scripts for creating and destroying assessment environments

This is necessary because the CloudWatch alarm resources associated
with the EC2 instances are created using for_each expressions that are
dynamic-ish.

jsf9k commented:
The lint GitHub Actions checks are currently failing due to psf/black#2964.  See also cisagov/skeleton-generic#105.

jsf9k commented:
cc: @BernardWoadzro, @mkreckel, @adevine31

dav3r commented on terraform_apply.sh:
```diff
@@ -0,0 +1,43 @@
+set -o pipefail
+
+# A script for creating assessment environments using
+# cisagov/cooll-assessment-terraform.
```
Typo:
```suggestion
# cisagov/cool-assessment-terraform.
```

dav3r commented on terraform_destroy_all.sh:
```diff
@@ -0,0 +1,28 @@
+set -o pipefail
+
+# A script for destroying assessment environments created using
+# cisagov/cooll-assessment-terraform.
```
Typo:
```suggestion
# cisagov/cool-assessment-terraform.
```

dav3r commented on terraform_apply.sh:
```diff
@@ -0,0 +1,43 @@
+
+# Perform a targeted apply to create the EC2 instances, then create
+# everything else.  This is a workaround for the dynamic-ish for_each
+# expressions mentinoed above.
```
Typo:
```suggestion
# expressions mentioned above.
```

dav3r commented on terraform_apply.sh:
```diff
@@ -0,0 +1,43 @@
+# Examples:
+# - See what would be created:
+#   $ AWS_PROFILE=cool-user AWS_SHARED_CREDENTIALS_FILE=~/.aws/production_credentials AWS_DEFAULT_REGION=us-east-1 ./terraform_apply.sh -var-file=envX-production.tfvars
+# - Destroy it!  (You will not be prompted.)
```
Copy/paste oversight:
```suggestion
# - Create it!  (You will not be prompted.)
```

jsf9k committed:
Fix typos

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Fix copy/paste oversight

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on terraform_apply.sh:
```diff
@@ -0,0 +1,43 @@
+# everything else.  This is a workaround for the dynamic-ish for_each
+# expressions mentioned above.
+terraform apply "${@}" \
+  -target=aws_iam_role_policy_attachment.provisionassessment_policy_attachment \
```
I always thought that we needed to first create the policy (`aws_iam_policy.provisionassessment_policy`) as well as the policy attachment.  If you only target the policy attachment, does that automatically target the policy too since it's a dependency?

jsf9k replied:
That definitely appeared to be what happened when I tested this code, and it makes sense to me that it would be the case.

dav3r replied:
That's nice to know, thanks.

dav3r commented on terraform_destroy_all.sh:
```diff
@@ -0,0 +1,28 @@
+
+# Create a list of resources to target from the terraform state and
+# then pass them to terraform destroy.
+terraform state list | sed "s/^/-target='/;s/$/'/" | xargs terraform destroy "$@"
```
It might be nice to expand either this comment or the one at the start of this file to add a little bit more detail about why a regular `terraform destroy` doesn't work in this case.

jsf9k replied:
I will make it so.

jsf9k replied:
Please see commit c4f7a5d.

dav3r replied:
Looks great, thanks for expounding on that.

dav3r reviewed (approved):
Makes sense, looks good, please see my one comment. 🚀

jsf9k committed:
Explain why the Terraform apply and destroy must be done in stages

dav3r commented on terraform_apply.sh:
```diff
@@ -0,0 +1,52 @@
+# are dynamic-ish.  When an untargeted apply is run, Terraform
+# verifies that each for_each attribute is computable without any
+# resources being instantiated.  That isn't possible in this case,
+# since Terraform must instantiate the EC2 instances before it can get
```
Typo:
```suggestion
# since Terraform must instantiate the EC2 instances before it can
```

dav3r commented on terraform_destroy_all.sh:
```diff
@@ -0,0 +1,33 @@
+# verifies that each for_each attribute is computable without any
+# resources needing to be instantiated.  That isn't possible in this
+# case, since Terraform must instantiate the EC2 instances before it
+# can get determine their IDs.  A targeted destroy avoids this check,
```
Typo:
```suggestion
# can determine their IDs.  A targeted destroy avoids this check,
```

jsf9k committed:
Fix typos

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Fix typo

jsf9k committed:
Ensure that all necessary resources are created before instantiating EC2 instances

In some cases the VPC endpoints, for example, were not functional
before the EC2 instances attempted to access them, and this should fix
that issue.

dav3r reviewed (approved):
👍 👍

felddy reviewed (approved):
This is quite a statement.  Well done.
I was going to mention `env -i` but thought better of it.  
Then mentioned it anyway I guess. 
💪

mcdonnnj commented on terraform_apply.sh:
```diff
@@ -0,0 +1,138 @@
+
+# Perform a targeted apply to create the EC2 instances, then create
+# everything else.  This is a workaround for the dynamic-ish for_each
+# expressions mentioned above.
```
I think this comment should better describe what's going on here. As it stands it describes the last two `terraform apply` calls in this chain of four.

jsf9k replied:
Please see commit 5264e7b.

mcdonnnj replied:
Perfect! Thanks!

mcdonnnj commented on terraform_apply.sh:
```diff
@@ -0,0 +1,138 @@
+    -target=aws_vpc_endpoint_subnet_association.ssm \
+    -target=aws_vpc_endpoint_subnet_association.ssmmessages \
+    -target=aws_vpc_endpoint_subnet_association.sts \
+    -target=null_resource.break_association_with_default_route_table \
+    -target=module.email_sending_domain_certreadrole \
+    -target=module.guacamole_certreadrole \
+    -target=module.read_terraform_state \
+    -target=module.run_shell_ssm_document \
+    -target=module.vpc_flow_logs \
+    -target=module.read_terraform_state \
```
Fixing the order and `module.read_terraform_state` is duplicated.

```suggestion
    -target=module.email_sending_domain_certreadrole \
    -target=module.guacamole_certreadrole \
    -target=module.read_terraform_state \
    -target=module.run_shell_ssm_document \
    -target=module.vpc_flow_logs \
    -target=null_resource.break_association_with_default_route_table \
```

mcdonnnj commented on terraform_apply.sh:
```diff
@@ -0,0 +1,138 @@
+    -target=aws_route.cool_operations \
+    -target=aws_route.cool_private \
+    -target=aws_route.external_operations \
+    -target=aws_route.external_private \
+    -target=aws_route53_vpc_association_authorization.assessment_private \
+    -target=aws_route_table_association.private_route_table_associations \
```
Are these sorted? I'm not a big fan of a number coming between symbols if they are.

jsf9k replied:
Please see commit d76d808.

mcdonnnj commented on terraform_destroy_all.sh:
```diff
@@ -0,0 +1,33 @@
+export AWS_SHARED_CREDENTIALS_FILE
+export AWS_DEFAULT_REGION
+
+# Create a list of resources to target from the terraform state and
```
It doesn't sound like you're referring to the `terraform state` call and more just Terraform's state. If I'm mistaken feel free to ignore this suggestion.

```suggestion
# Create a list of resources to target from the Terraform state and
```

mcdonnnj reviewed (approved):
Makes sense. I have some quibbles and suggestions for your consideration.

jsf9k committed:
Alphabetize target resources

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Capitalize Terraform

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Improve accuracy of comment contents

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Improve sorting of targeted resources

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
